src/mixins/object_interactivity.mixin.ts

```diff
@@ -1,394 +1,497 @@
-          // width and height of the above function are the size of the bbox.
-          if (
-            control.withConnection &&
-            control.getVisibility(fabricObject, key)
-          ) {
-            // reset movement for each control
-            shouldStroke = true;
-            ctx.moveTo(control.x * size.x, control.y * size.y);
-            ctx.lineTo(
-              control.x * size.x + control.offsetX,
-              control.y * size.y + control.offsetY
-            );
-          }
-        });
-        shouldStroke && ctx.stroke();
-
-        return this;
-      },
-
-      /**
-       * Draws corners of an object's bounding box.
-       * Requires public properties: width, height
-       * Requires public options: cornerSize, padding
-       * @param {CanvasRenderingContext2D} ctx Context to draw on
-       * @param {Object} styleOverride object to override the object style
-       * @return {fabric.Object} thisArg
-       * @chainable
-       */
-      drawControls: function (ctx, styleOverride) {
-        styleOverride = styleOverride || {};
-        ctx.save();
-        var retinaScaling = this.canvas.getRetinaScaling(),
-          p;
-        ctx.setTransform(retinaScaling, 0, 0, retinaScaling, 0, 0);
-        ctx.strokeStyle = ctx.fillStyle =
-          styleOverride.cornerColor || this.cornerColor;
-        if (!this.transparentCorners) {
-          ctx.strokeStyle =
-            styleOverride.cornerStrokeColor || this.cornerStrokeColor;
-        }
-        this._setLineDash(
-          ctx,
-          styleOverride.cornerDashArray || this.cornerDashArray
-        );
-        this.setCoords();
-        this.forEachControl(function (control, key, fabricObject) {
-          if (control.getVisibility(fabricObject, key)) {
-            p = fabricObject.oCoords[key];
-            control.render(ctx, p.x, p.y, styleOverride, fabricObject);
-          }
+type TOCoord = IPoint & {
+  corner: TCornerPoint;
+  touchCorner: TCornerPoint;
+};
+
+export class InteractiveFabricObject extends FabricObject {
+
+
+  /**
+   * Describe object's corner position in canvas element coordinates.
+   * properties are depending on control keys and padding the main controls.
+   * each property is an object with x, y and corner.
+   * The `corner` property contains in a similar manner the 4 points of the
+   * interactive area of the corner.
+   * The coordinates depends from the controls positionHandler and are used
+   * to draw and locate controls
+   * @memberOf fabric.Object.prototype
+   */
+  oCoords: Record<string, TOCoord> = {};
+
+  /**
+   * keeps the value of the last hovered corner during mouse move.
+   * 0 is no corner, or 'mt', 'ml', 'mtr' etc..
+   * It should be private, but there is no harm in using it as
+   * a read-only property.
+   * this isn't cleaned automatically. Non selected objects may have wrong values
+   * @type number|string|any
+   * @default 0
+   */
+  __corner: number | string;
```

Comment:
why not make all 0 assignments null or undefined?

Reply:
yes i thought the same, we could move it to 0 as default rather than null. I don't think anyone is gonna cry for that.

Reply:
I am confused
I thought removing 0 in favor of null/undefined

Reply:
sorry nor zero nor null. just empty string or optional with undefined.

i din't like null it often makes no sense to me.

Reply:
I am asking since I saw the code still has `__corner = 0`

Reply:
either empty string or undefined i suppose. With optional string ( and so undefined ) looking more clean